lib/datadog/core/telemetry/collector.rb

```diff
@@ -0,0 +1,207 @@
+          end
+        end
+
+        # Forms a telemetry host object
+        def host
+          Telemetry::V1::Host
+            .new(
+              container_id: Core::Environment::Container.container_id,
+              hostname: Core::Environment::Platform.hostname,
+              kernel_name: Core::Environment::Platform.kernel_name,
+              kernel_release: Core::Environment::Platform.kernel_release,
+              kernel_version: Core::Environment::Platform.kernel_version
+            )
+        end
+
+        # Forms a telemetry app-started integrations object
+        def integrations
+          Datadog.registry.map do |integration|
+            is_instrumented = instrumented?(integration)
+            is_enabled = is_instrumented && patched?(integration)
+            Telemetry::V1::Integration
+              .new(
+                name: integration.name.to_s,
+                enabled: is_enabled,
+                version: integration_version(integration),
+                compatible: integration_compatible?(integration),
+                error: (integration_error(integration) if is_instrumented && !is_enabled),
+                auto_enabled: is_enabled ? integration_auto_instrument?(integration) : nil
+              )
+          end
+        end
+
+        # Returns the runtime ID of the current process
+        def runtime_id
+          Datadog::Core::Environment::Identity.id
+        end
+
+        # Returns the current as a UNIX timestamp in seconds
+        def tracer_time
+          Time.now.to_i
+        end
+
+        private
+
+        def env
+          Datadog.configuration.env
+        end
+
+        def service_name
+          Datadog.configuration.service
+        end
+
+        def service_version
+          Datadog.configuration.version
+        end
+
+        def tracer_version
+          Core::Environment::Identity.tracer_version
+        end
+
+        def products
+          profiler_obj = profiler
+          appsec_obj = appsec
+          profiler_obj || appsec_obj ? Telemetry::V1::Product.new(profiler: profiler_obj, appsec: appsec_obj) : nil
+        end
+
+        def profiler
+          version = profiler_version
+          Telemetry::V1::Profiler.new(version: version) if version
+        end
+
+        def profiler_version
+          tracer_version if Datadog.configuration.respond_to?(:profiling) && Datadog.configuration.profiling.enabled
+        end
+
+        def appsec
+          version = appsec_version
+          Telemetry::V1::AppSec.new(version: version) if version
+        end
+
+        def appsec_version
+          tracer_version if Datadog.configuration.respond_to?(:appsec) && Datadog.configuration.appsec.enabled
+        end
+
+        def configuration_variables
+          configurations = []
+          environment_collector = Core::Diagnostics::EnvironmentCollector.new
+          [
+            Telemetry::V1::Configuration.new(name: 'diagnostics.debug', value: environment_collector.debug),
```

Comment:
_Off-topic comment_
It seems to me that `Telemetry::V1::Configuration` doesn't need its own class.
A list of `Telemetry::V1::Configuration`s is simply a Ruby Hash.

Reply:
This doesn't need to be address here in this PR, but I think we should simplify our data model by removing Telemetry::V1::Configuration.

Reply:
Tempted to agree. The upsides to doing this, however, is that 1) hashes take more memory, and 2) typing this object makes it easier to serialize later.